### PR TITLE
enum WIP

### PR DESCRIPTION
**Pull request by stillinbeta**
Thank you for contributing to pyo3!

Here are some things you should check for submitting your pull request:

 - Run `cargo fmt` (This is checked by travis ci)
 - Run `cargo clippy` and check there are no hard errors (There are a bunch of existing warnings; This is also checked by travis)
 - If applicable, add an entry in the changelog.
 - If applicable, add documentation to all new items and extend the guide.
 - If applicable, add tests for all new or fixed functions
 - If you changed any python code, run `black .`. You can install black with `pip install black`)

You might want to run `tox` (`pip install tox`) locally to check compatibility with all supported python versions. If you're using linux or mac you might find the Makefile helpful for testing.

### CODE REVIEW

**Comment by davidhewitt** [comment]
@stillinbeta many thanks for kicking this off, I'm thrilled to see this! With respect to your comment on the other thread:

> I made an attempt (#1045), but I got frustrated by the amount of boilerplate and traits required.

If we were to work together to design this functionality, have you got interest in continuing onwards with this PR?

If you've lost interest that's also fine, I'll pick this up eventually when I get round to thinking about this feature harder.

Reading through what you've started here, I have a few initial thoughts:

- The variant parsing code you've written looks clean and straightforward for these "C-style" enums, which is nice to see. 👍 
- I'm wondering whether we need to implement `PyClass` and all its gubbins for `#[pyenum]`, or whether we should go down a different direction.
  - I don't _think_ that we'll need either inheritance or being able to have `&mut self` receivers in `#[pymethods]` for these "C-style" enums. These are two of the main reasons `PyClass` is so messy, so maybe we can design something leaner here.
  - On the other hand, at the moment the `#[pyproto]` implementations are _heavily_ coupled to `PyClass`. That's not necessarily an issue, but it does mean that if we want to use `#[pyproto]` with `#[pyenum]` then we're either forced to use `PyClass`, or we'll have to do some refactoring...
- It would likely be nice to add each enum variant as a `#[classattr]` automatically.

**Comment by kngwyu** [comment]
Thanks!

> I don't think that we'll need either inheritance or being able to have &mut self receivers in #[pymethods] for these "C-style" enums.

Yeah, but I don't think both inheritance and `&mut self` don't contribute the code complexity here.
I think what we need is just to use the common code as pyclass.

**Comment by stillinbeta** [comment]
My initial intent was to inherit from enum, but that blocked on #991. I
agree that modularizing the pyclass macro is probably the way forward,
since we're definitely going to need some of it (`EnumName.EnumVariant` was
the UX I was going for, a la the enum modules, and I think that basically
necessitates a class)

**Comment by davidhewitt** [comment]
Agreed. Perhaps if I drafted up a solution to #991, we could get that merged which would pave the way for inheriting from `enum.Enum` in this PR?

I guess we can continue building out the other elements of this PR in the meanwhile!

> `EnumName.EnumVariant` was the UX I was going for, a la the enum modules, and I think that basically necessitates a class

If you haven't seen it yet, you should check out the `#[classattr]` annotation we support in `#[pymethods]` blocks. We already went through a lot of trouble to allow that to safely support attributes that are instances of the containing class, so you should definitely be able to re-use that to expose the variants.

**Comment by stillinbeta** [comment]
I took a stab at getting rid of all the PyClass constraint, but it's going to require a slot of boilerplate on its own:

We need a Layout, and the only blanket implementation is for PyClass: https://docs.rs/pyo3/0.11.1/pyo3/type_object/trait.PyLayout.html#impl-PyLayout%3CT%3E
So does Initializer, which is only implemented in for PyClass and native types (which this definitely isn't). https://docs.rs/pyo3/0.11.1/pyo3/pyclass_init/struct.PyClassInitializer.html

I managed to get enough of the traits in there to compile (was just missing one, it turns out), but it feels... janky. We don't actually need the Inventory or the ProtoRegistry, but the constraints on PyClass mandate them. Do we think we could loosen some of those constraints, or would that break other things?

**Comment by davidhewitt** [comment]
Thanks for continuing to iterate on this!

Regarding the layout, I think that we could try adding a new struct like this:

```
struct PyEnumLayout<T> {
    base: PyAny,
    value: T
}
```

This new type could get a new blanket layout?

As for the initializer, I think that we can use `PyClassInitializer` by adding a new function similar to `create_cell_from_subtype`. The new function would instead return `PyResult<*mut PyEnumLayout<T>>`.

> We don't actually need the Inventory or the ProtoRegistry

I actually think we probably do want to support `#[pymethods]` and `#[pyproto]` for enum classes? So I think we do actually need the inventories.

**Comment by kngwyu** [comment]
> I took a stab at getting rid of all the PyClass constraint, but it's going to require a slot of boilerplate on its own:

I **really** don't think we should do so.
So now we're going to introduce new lots of boilerplates for reducing boilerplates for pyclass?
I don't think it's so meaningful and believe that reusing all `pyclass` components is the simplest way and also easy to maintain.

**Comment by kngwyu** [comment]
I think what we really need here is to provide a nice API for *our* Rustic enum. How to implement it does not matter so much. So I recommend to implement it just as pyclass.

**Comment by gilescope** [comment]
I keep wondering if another way to think about enums is if we supported static getters on pyclasses? That would look from the python side very similar esp if we could subclass Enum...

**Comment by stillinbeta** [comment]
> I actually think we probably do want to support #[pymethods] and #[pyproto] for enum classes? So I think we do actually need the inventories.

I'm not sure I agree with this. Do we have examples of Enums in python with methods on them? It seems like usually they're just sigils in Python.

**Comment by programmerjake** [comment]
> I'm not sure I agree with this. Do we have examples of Enums in python with methods on them? It seems like usually they're just sigils in Python.

Example:
https://git.libre-soc.org/?p=ieee754fpu.git;a=blob;f=src/ieee754/div_rem_sqrt_rsqrt/core.py;h=9f925b871493a6885d92366d648ea6c57bb39c20;hb=HEAD#l28
```python3
class DivPipeCoreOperation(enum.Enum):
    """ Operation for ``DivPipeCore``.

    :attribute UDivRem: unsigned divide/remainder.
    :attribute SqrtRem: square-root/remainder.
    :attribute RSqrtRem: reciprocal-square-root/remainder.
    """

    SqrtRem = 0
    UDivRem = 1
    RSqrtRem = 2

    def __int__(self):
        """ Convert to int. """
        return self.value

    @classmethod
    def create_signal(cls, *, src_loc_at=0, **kwargs):
        """ Create a signal that can contain a ``DivPipeCoreOperation``. """
        return Signal(range(min(map(int, cls)), max(map(int, cls)) + 2),
                      src_loc_at=(src_loc_at + 1),
                      decoder=lambda v: str(cls(v)),
                      **kwargs)
```

**Comment by stillinbeta** [comment]
oh yep, fair enough. In that case, I think PyClass is the way to go. Though I think perhaps initially I'll try to land the change without pymethod support, then we can add it.

**Comment by gilescope** [comment]
Any luck @stillinbeta ?

**Comment by stillinbeta** [comment]
Sorry, I don't have a lot of time during the week to work on this (day job
drains my programming energy). Will give it a shot tomorrow!

**Comment by kngwyu** [comment]
Don't worry, that’s the order of OSS

**Comment by stillinbeta** [comment]
Made some progress this weekend! You can see that you can now return enums from pyfunctions, and they'll be Python-ified properly. However, I ran into issues when trying to pass one back:

```
error[E0119]: conflicting implementations of trait `pyo3::FromPyObject<'_>` for type `MyEnum`:
 --> tests/test_enum.rs:6:1
  |
6 | #[pyenum]
  | ^^^^^^^^^
  |
  = note: conflicting implementation in crate `pyo3`:
          - impl<'a, T> pyo3::FromPyObject<'a> for T
            where T: pyo3::PyClass, T: std::clone::Clone;
  = note: this error originates in an attribute macro (in Nightly builds, run with -Z macro-backtrace for more info)
```

Any ideas how to get around this? Should be implementing a different trait, like `PyTryInto`?

File: pyo3-derive-backend/src/pyenum.rs [hunk 1]
```diff
@@ -0,0 +1,214 @@
...
+    variants: &Vec<(syn::Ident, syn::ExprLit)>,
+) -> syn::Result<TokenStream> {
+    let matches = variants
+        .iter()
+        .map(|(ident, _)| {
+            variant_enumname(enum_, ident).map(|cls| {
+                let name = cls.to_string();
+                // TODO: this should be
+                // #cls as pyo3::type_object::PyTypeObject>::type_object(py).compare()
+                // but I can't figure out how to get py inside extract()
```
Comment (kngwyu) [inline pyo3-derive-backend/src/pyenum.rs:89]: You can use `pyo3::PyNativeType::py(ob)`

**Comment by kngwyu** [comment]
How does this work without `impl FromPyObject`?

**Comment by stillinbeta** [comment]
Do you mean “does it compile without that part?”

If so, yes, and the appropriate test cases pass

**Comment by gilescope** [comment]
Nice work @stillinbeta ! As an extension I wonder if it would be possible to use enums where the numeric value is not specified? I know on the rust side you could access the https://doc.rust-lang.org/std/mem/fn.discriminant.html but whether that helps at all I'm less certain.

**Comment by kngwyu** [comment]
> If so, yes, and the appropriate test cases pass

Then why should we need that impl?

File: pyo3-derive-backend/src/pyenum.rs [hunk 2]
```diff
@@ -0,0 +1,214 @@
...
+            type BaseNativeType = pyo3::PyAny;
+        }
+
+        #protoregistry
+
+        #extractext
+
+        impl pyo3::pyclass::PyClassAlloc for #typ {}
+
+        // TODO: handle not in send
```
Comment (davidhewitt) [inline pyo3-derive-backend/src/pyenum.rs:207]: I think it's reasonable to expect `#[pyenum]` to always be `Send`? It wouldn't ever carry nontrivial data afaik.

**Comment by davidhewitt** [comment]
> Then why should we need that impl?

+1. Like for `#[pyclass]`, I think this `impl` gets derived automatically if you have `#[derive(Clone)]`.

**Comment by stillinbeta** [comment]
The impl is because I want passing the class in to rust to turn into the
enum variant, not the a type object. I don't think that'll work
automatically

File: tests/test_enum.rs [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+    let f = wrap_pyfunction!(return_enum)(py);
+    let mynum = py.get_type::<MyEnum>();
+
+    py_run!(py, f mynum, "assert f() == mynum.Variant")
+}
+
+#[pyfunction]
+fn enum_arg(e: MyEnum) {
+    assert_eq!(MyEnum::OtherVariant, e)
+}
```
Comment (davidhewitt) [inline tests/test_enum.rs:40]: I suggest making this 

```suggestion
fn enum_arg(e: MyEnum) -> String {
    format!("{}", e)
}
```

File: tests/test_enum.rs [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+
+#[test]
+fn test_enum_arg() {
+    let gil = Python::acquire_gil();
+    let py = gil.python();
+    let f = wrap_pyfunction!(enum_arg)(py);
+    let mynum = py.get_type::<MyEnum>();
+
+    py_run!(py, f mynum, "f(mynum.Variant)")
```
Comment (davidhewitt) [inline tests/test_enum.rs:49]: If you take my suggestion from above, we can test both cases:

```suggestion
    py_run!(py, f mynum, "assert f(mynum.Variant) == 'Variant'")
    py_run!(py, f mynum, "assert f(mynum.OtherVariant) == 'OtherVariant'")
```

**Comment by stillinbeta** [comment]
the `test_enum_arg` behaviour doesn't work right now, that's the problem: 
```
Traceback (most recent call last):
  File "<string>", line 1, in <module>
TypeError: Can't convert <class 'Variant'> to MyEnum
test test_enum_arg ... FAILED
```

There's no way to automatically determine that I want the class Variant to map to MyEnum::Variant, I need to produce that code myself.

File: pyo3-derive-backend/src/pyenum.rs [hunk 5]
```diff
@@ -0,0 +1,214 @@
...
+
+        #enum_cls
+
+        #(
+            struct #variant_names;
+        )*
+
+        #(
+            #variant_cls
+        )*
```
Comment (davidhewitt) [inline pyo3-derive-backend/src/pyenum.rs:62]: I don't think this bit is needed - instead you should just be making the variants be instances of the the enum class. (Like how it is with the `enum` module.)

**Comment by davidhewitt** [comment]
Oh, I see what you mean now! I don't think that `MyClass.Variant` should be a separate type? I've marked that code in the comment above.

**Comment by stillinbeta** [comment]
Ah yeah, that makes sense. I misinterpreted how Enum was architected.

**Comment by stillinbeta** [comment]
The one advantage of every variant having its own class is it made mapping the variants to their classes fit better into the model of struct -> class. I think to create instances of Variant, I need to make some use of IntoPython, any ideas what that would look like? I'm a little lost in the type system right now.

File: pyo3-derive-backend/src/pyenum.rs [hunk 6]
```diff
@@ -0,0 +1,214 @@
...
+            variant_enumname(enum_, ident).map(|cls| {
+                quote! {
+                    #enum_::#ident => <#cls as pyo3::type_object::PyTypeObject>::type_object(py).to_object(py),
+                }
+            })
+        })
+        .collect::<syn::Result<Vec<_>>>()?;
+
+    Ok(quote! {
+        impl pyo3::FromPy<#enum_> for pyo3::PyObject {
```
Comment (davidhewitt) [inline pyo3-derive-backend/src/pyenum.rs:132]: With the rebase on master, you'll need to change this to `impl IntoPy<PyObject> for #enum`

**Comment by davidhewitt** [comment]
I think `MyClass::Variant.into_py(py)` should already do the into-python conversion you're asking for?

FWIW if you rebase on master I just merged some simplifications to the type system, which hopefully will help!

**Comment by stillinbeta** [comment]
```rust
impl pyo3::IntoPy<pyo3::PyObject> for #enum_ {
            fn into_py(self, py: pyo3::Python) -> pyo3::PyObject {
                self.into_py(py)
            }
        }
```

is definitely not going to do it, that's just recursion.

**Comment by davidhewitt** [comment]
Oh heh I see what you're asking 😄

You probably want to use the same implementation as `#[pyclass]`, which uses `Py::new`

https://github.com/PyO3/pyo3/blob/bcb90775b43659a58f44755d3fbaf92ac493d8c2/pyo3-derive-backend/src/pyclass.rs#L379

File: pyo3-derive-backend/src/pyenum.rs (hunk 2 above)
Comment (gilescope) [reply pyo3-derive-backend/src/pyenum.rs:207]: Maybe go further and have pyenum always be 'Copy' as we are doing small data here rather than ADTs?

Comment (davidhewitt) [reply pyo3-derive-backend/src/pyenum.rs:207]: Perhaps, though I'm not sure it's our place to add `#[derive(Copy, Clone)]` automatically to the struct.

**Comment by gilescope** [comment]
Given https://github.com/PyO3/pyo3/pull/1065 is merged in, do we want to also try and get this landed so they can be in 0.12 together or does that PR also make it possible to convert simple rust enums to python enums so this PR is now not needed?

**Comment by davidhewitt** [comment]
#1065 enables the Python -> Rust direction, but it doesn't really solve the case of data enums like we see here (afaik).

I'm now thinking that this new macro `#[pyenum]` makes sense to land in 0.13, with `#[pyexception]`. That should also give time for this to be finished off.

**Comment by gilescope** [comment]
Still keen on using this and having proper python enums! What's left to do?

**Comment by hwchen** [comment]
Same here, I'm definitely looking forward to this and I'm happy to help.

**Comment by davidhewitt** [comment]
I think at this point it's reasonable to assume the original PR author has unfortunately not been able to find the time to finish this off, so I guess they wouldn't mind if anyone is interested in copying this branch and reopening it as a new PR.

I think that rebasing it on master, applying the review changes I've made here, and then fixing any subsequent test failures would be a great starting point. I don't think there's particularly far to go for us to merge this!

**Comment by stillinbeta** [comment]
Yeah, I've mostly moved onto other projects, sorry for not un-licking this
particular cookie sooner. Go forth with my blessing, and I'll try to answer
questions about my code if they come up.

**Comment by aviramha** [comment]
Hey, I'll take this.

**Comment by davidhewitt** [comment]
This is now superseded by #2002. Thanks again for kicking this off.

**Pull request closed without merging**